### ISSUE TITLE
Docs: add allowance in no-explicit-any docs that any is sometimes a necessary last resort

### ISSUE THREAD

**Issue by ericelliott** [typescript-eslint/typescript-eslint#7354]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

The current documentation reads as if explicit any is almost always an error that needs fixing, but due to a lack of TypeScript features (e.g. higher-kinded types or Haskell-style custom infix operators) explicit any is often the only correct fix for typing higher-order functions.

e.g.:

```TypeScript
// eslint-disable-next-line no-explicit-any
type a2a = (x: any) => any;
type compose = (...fns: a2a[]) => a2a;
const pipe: compose = (...fns) => x => fns.reduce((y, f) => f(y), x);
const compose: compose = (...fns) => x => fns.reduceRight((y, f) => f(y), x);

type n2n = (n: number) => number;
const g: n2n = n => n + 1;
const f: n2n = n => n * 2;

const h: n2n = pipe(g, f);
const j: n2n = compose(f, g);
```

### Affected URL(s)

https://typescript-eslint.io/rules/no-explicit-any/

**Comment by JoshuaKGoldberg** [comment]
👋 thanks for sending an issue & quick PR to fix it! Reading through this & #7355, I think this might be more of an _accuracy_ issue in the docs than _confusion_. There are two points I think the docs might want to improve on:

* Mentioning `unknown` or writing actual types as a preference: tracked in https://github.com/typescript-eslint/typescript-eslint/issues/7136; definitely a point of improvement
* Mentioning the edge cases in the type system that force `unknown` to not be usable as an alternative

For the latter, we can take a PR. Thoughts?

**Comment by bradzacher** [comment]
> The current documentation reads as if explicit any is almost always an error that needs fixing

I think as a general statement "don't ever use `any`" is a really good piece of advice. 
There are truly very, very few cases where you actually want `any` as your type.

Why is it good advice to give?
Many TS novices will reach for `any` because it's there - they don't know how to solve a problem so they just throw in `any` and forget about the hole they just put in their code. Even advanced users do this - ESP engineers working at companies - writing "working" code is more important than writing type-safe code - so they'll throw an `any` in to avoid a problem so they can satisfy CI and ship their feature.

There *are* definitely cases where an `any` can be handy. [We ourselves have a small handful of them in our codebase!](https://github.com/search?q=repo%3Atypescript-eslint%2Ftypescript-eslint+%2F%5Cb%28%3A+%7C+as+%7C+%3D%3E+%29any%5Cb%2F+NOT+path%3A%2F%5C%2Fdocs%5C%2F%2F+NOT+path%3A%2F.%28mdx%3F%7C.test.tsx%3F%7Cpatch%29%24%2F+NOT+path%3A%2F%5C%2F%28mock%7Cfixture%29s%3F%5C%2F%2F++NOT+path%3A%2F%5C.github%2F+NOT+path%3A%2Fintegration-tests%2F&type=code) But that doesn't mean that taking the stance of "get rid of the `any`" is wrong.

As an example - in your snippet above you stated that the `any` was the only way to get the correct result from your code - [yet using a few generics you can achieve the same result with fewer "override the `any`" type annotations](https://www.typescriptlang.org/play?#code/C4TwDgpgBAhgTDAPAFQHxQLxQBQA8BcUyAlJusgNwBQokUAxgPYC2YjAztFiqtgHQCAZgDt2heEjQBtALqkM6CT2pNRwKGACWkQk1YcuOAXxHt56XGSim+AJwgATAK70I2bCAA0189Y-FvXGIVRjUGFjZOXQiDTCMhUV9LBWtRO0cXCAAlTQBzAAtgdy8fK0F-QOCqKlV2dVy47GFCYSdmACMIW19hKABqKABGELDBRuaoVo6unqgAKig4ahrQuqh8uK1IbFzvQSqAegOoE6gAPQB+FbCAKzi9SLdBb1zD49PL6qA). I'm sure your example was a much simplified example of real code as I know there are definitely cases where trying to add generics or explicitly type things becomes way too difficult.
As an aside - for those cases [we generally prefer reaching for `@ts-expect-error` instead](https://github.com/search?q=repo%3Atypescript-eslint%2Ftypescript-eslint+NOT+path%3A%2F%5C%2Fdocs%5C%2F%2F+NOT+path%3A%2F.%28mdx%3F%7C.test.tsx%3F%7Cpatch%29%24%2F+NOT+path%3A%2F%5C%2F%28mock%7Cfixture%29s%3F%5C%2F%2F++NOT+path%3A%2F%5C.github%2F+NOT+path%3A%2Fintegration-tests%2F+%2F%40ts-%28expect-error%7Cignore%29%2F&type=code) as it's generally got a smaller blast radius for cases that you need to "work around" the typechecker.

The important thing to note that `any` is 100% a dangerous thing to use in code. Your example is only safe purely because you've explicitly annotated the variables to "hide" the `any`. Essentially you've allowed the unsafety and covered it up quickly. But that code is very easy to get wrong!! For example imagine if your library provided an `n2s` type and someone accidentally typed `h: n2s` instead of `h: n2n` - that's all fine until someone accidentally does `h(1).includes(string)` which typechecks fine but blows up at runtime!

---

With all that being said - I want to reiterate that the stance of "don't ever use `any`" is still a good stance to have and I don't think that the docs are wrong for painting `any` as an error to be fixed.

**Comment by ericelliott** [comment]
> As an example - in your snippet above you stated that the any was the only way to get the correct result from your code - [yet using a few generics you can achieve the same result with fewer "override the any" type annotations](https://www.typescriptlang.org/play?#code/C4TwDgpgBAhgTDAPAFQHxQLxQBQA8BcUyAlJusgNwBQokUAxgPYC2YjAztFiqtgHQCAZgDt2heEjQBtALqkM6CT2pNRwKGACWkQk1YcuOAXxHt56XGSim+AJwgATAK70I2bCAA0189Y-FvXGIVRjUGFjZOXQiDTCMhUV9LBWtRO0cXCAAlTQBzAAtgdy8fK0F-QOCqKlV2dVy47GFCYSdmACMIW19hKABqKABGELDBRuaoVo6unqgAKig4ahrQuqh8uK1IbFzvQSqAegOoE6gAPQB+FbCAKzi9SLdBb1zD49PL6qA).

This would be a stronger point if the  proposed alternative was a valid type for compose or pipe, which can take any number of *differently typed* functions: a => b, b => c to produce a => c. Sadly, it's not. Throw a number => string on the end and the generic typed alternative blows up.

If you prefer `// @ts-expect-error -- because TypeScript has poor support for higher-order functions`  that's fine, too, but this still feels like a very valid use-case for `any`.

I like having good types in my programs to, but until TypeScript can do this stuff properly, more like Haskell, we sometimes need the escape hatch (as you demonstrated with MANY escape hatch uses). I'm still reeling from the whiplash of dogmatic reply paired with laissez faire escape hatch code examples - some of which actually have good well-typed alternatives, unlike this example. 🤣

![IMG_9579](https://github.com/typescript-eslint/typescript-eslint/assets/364727/85248291-0dc5-4da5-a34f-18ebfabd88ad)

**Comment by ericelliott** [comment]
> Mentioning the edge cases in the type system that force unknown to not be usable as an alternative

This appears to be one of those cases - tack on a function to be composed which goes from b => c instead of a => b, and `unknown` complains in this context.

**Comment by ericelliott** [comment]
I believe all the conversations in PR https://github.com/typescript-eslint/typescript-eslint/pull/7355 have been resolved.
